dbms/src/DataStreams/WindowBlockInputStream.cpp

```diff
@@ -272,12 +274,120 @@ Int64 WindowTransformAction::getPartitionEndRow(size_t block_rows)
     return left;
 }
 
+RowNumber WindowTransformAction::stepForward(const RowNumber & current_row, UInt64 n)
+{
+    auto dist = distance(current_row, partition_start);
+    assert(dist >= 0);
```

Comment:
So the return type of `distance` can be `UInt64` too?

Reply:
> So the return type of `distance` can be `UInt64` too?

Although this function returns positive number now, but I can't ensure if this behaviour can always be guaranteed, so I think negative number is better.